File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Power/AbstractPower.ts
import { IGameAction } from "../../../../Core/Module/Event/IGameAction";
import { RemoveSpecificPowerAction } from "../Action/Common/RemoveSpecificPowerAction";
import DungeonManager from "../DungeonManager";
import EffectKit from "../Effect/EffectKit";
import { FlashPowerEffect } from "../Effect/FlashPowerEffect";
import { GainPowerEffect } from "../Effect/GainPowerEffect";
import { AtEndOfTurnMessage } from "../Events/AtEndOfTurnMessage";
import { AtEndOfTurnPreEndTurnCardsMessage } from "../Events/AtEndOfTurnPreEndTurnCardsMessage";
import { AtStartOfTurnMessage } from "../Events/AtStartOfTurnMessage";
import { CalcDamageValueMessage } from "../Events/CalcDamageValueMessage";
import { CalcFinalDamageValueMessage } from "../Events/CalcFinalDamageValueMessage";
import { CalcFinalHurtValueMessage } from "../Events/CalcFinalHurtValueMessage";
import { CalcHurtValueMessage } from "../Events/CalcHurtValueMessage";
import { OnDamageAfter } from "../Events/OnDamageAfter";
import { OnDamageBefore } from "../Events/OnDamageBefore";
import { OnDeathMessage } from "../Events/OnDeathMessage";
import { OnHurt } from "../Events/OnHurt";
import { OnHurtAfter } from "../Events/OnHurtAfter";
import { OnHurtBefore } from "../Events/OnHurtBefore";
import { OnPreUseCardMessage } from "../Events/OnPreUseCardMessage";
import { PowerModifiedAmountMessage } from "../Events/PowerModifiedAmountMessage";
import { PreModifyBlockMessage } from "../Events/PreModifyBlockMessage";
import AbstractCreature from "../Unit/AbstractCreature";


export abstract class AbstractPower {

    public abstract Id:string;
    public abstract Name: string;
    //获得描述
    public abstract GetDescription(...args: any[]): string;
    //图标
    public abstract Icon:string;
    //层数
    private _Amount : number;
    //可否叠加
    public CanStack:boolean;
    //可否驱散
    public CanRemove : boolean;
    //结算优先级
    public Priority : number;
    //该Buff的拥有者
    public Owner: AbstractCreature;
    //该Buff的施法来源
    public Source: AbstractCreature;
    //能力类型(Buff,Debuff)
    public abstract get Type() : PowerType;
    get Amount(): number {
        return this._Amount;
    }

    set Amount(value: number) {
        let preAmount = this._Amount;
        this._Amount = value;
        this.OnModified(preAmount,value);
        DungeonManager.MessageManager.Send(PowerModifiedAmountMessage.Id,new PowerModifiedAmountMessage(this,preAmount,value));
    }

    //重载该方法需要调用super.OnInit();
    public OnInit() {
        DungeonManager.MessageManager.Add(OnHurtBefore.Id,this.OnHurtBefore)
        DungeonManager.MessageManager.Add(OnHurt.Id,this.OnHurt)
        DungeonManager.MessageManager.Add(OnHurtAfter.Id,this.OnHurtAfter)
        DungeonManager.MessageManager.Add(AtEndOfTurnMessage.Id,this.AtEndOfTurn);
        DungeonManager.MessageManager.Add(AtStartOfTurnMessage.Id,this.AtStartOfTurn);
        DungeonManager.MessageManager.Add(CalcDamageValueMessage.Id,this.CalcDamageValue);
        DungeonManager.MessageManager.Add(OnDamageBefore.Id,this.OnDamageBefore);
        DungeonManager.MessageManager.Add(OnDamageAfter.Id,this.OnDamageAfter);
        DungeonManager.MessageManager.Add(CalcFinalDamageValueMessage.Id,this.CalcFinalDamageValue);
        DungeonManager.MessageManager.Add(CalcHurtValueMessage.Id,this.CalcHurtValue);
        DungeonManager.MessageManager.Add(CalcFinalHurtValueMessage.Id,this.CalcFinalHurtValue);
        DungeonManager.MessageManager.Add(OnDeathMessage.Id,this.OnDeath);
        DungeonManager.MessageManager.Add(AtEndOfTurnPreEndTurnCardsMessage.Id,this.AtEndOfTurnPreEndTurnCards);
        DungeonManager.MessageManager.Add(PreModifyBlockMessage.Id,this.PreModifyBlock);
    }
    
    //重载该方法需要调用super.OnRemove();
    public OnRemove(){
        DungeonManager.MessageManager.Remove(OnHurtBefore.Id,this.OnHurtBefore)
        DungeonManager.MessageManager.Remove(OnHurt.Id,this.OnHurt)
        DungeonManager.MessageManager.Remove(OnHurtAfter.Id,this.OnHurtAfter)
        DungeonManager.MessageManager.Remove(AtEndOfTurnMessage.Id,this.AtEndOfTurn);
        DungeonManager.MessageManager.Remove(AtStartOfTurnMessage.Id,this.AtStartOfTurn);
        DungeonManager.MessageManager.Remove(CalcDamageValueMessage.Id,this.CalcDamageValue);
        DungeonManager.MessageManager.Remove(OnDamageBefore.Id,this.OnDamageBefore);
        DungeonManager.MessageManager.Remove(OnDamageAfter.Id,this.OnDamageAfter);
        DungeonManager.MessageManager.Remove(CalcFinalDamageValueMessage.Id,this.CalcFinalDamageValue);
        DungeonManager.MessageManager.Remove(OnDamageAfter.Id,this.OnDamageAfter);
        DungeonManager.MessageManager.Remove(CalcFinalDamageValueMessage.Id,this.CalcFinalDamageValue);
        DungeonManager.MessageManager.Remove(CalcHurtValueMessage.Id,this.CalcHurtValue);
        DungeonManager.MessageManager.Remove(CalcFinalHurtValueMessage.Id,this.CalcFinalHurtValue);
        DungeonManager.MessageManager.Remove(OnDeathMessage.Id,this.OnDeath);
        DungeonManager.MessageManager.Remove(AtEndOfTurnPreEndTurnCardsMessage.Id,this.AtEndOfTurnPreEndTurnCards);
        DungeonManager.MessageManager.Remove(PreModifyBlockMessage.Id,this.PreModifyBlock);
    }
    
    public PlayApplyPowerSfx(){}
    protected AddToBot(action : IGameAction){DungeonManager.ActionManager.AddToBottom(action);}
    protected AddToTop(action : IGameAction){DungeonManager.ActionManager.AddToTop(action);}
    public Flash(silent:boolean = true){
        EffectKit.Inst().Play(new GainPowerEffect(this,silent));
        EffectKit.Inst().Play(new FlashPowerEffect(this));
    }
    //该方法仅用于计算伤害.最终伤害由OnHurt结算.
    protected CalcHurtValue(msg:CalcDamageValueMessage){}
    //该方法仅用于计算伤害.最终伤害由OnHurt结算.
    protected CalcFinalHurtValue(msg:CalcDamageValueMessage){}
    //当玩家受到伤害之前
    protected OnHurtBefore(msg:OnHurtBefore){}
    //当玩家受到伤害时
    protected OnHurt(msg:OnHurtBefore){}
    //当玩家受到伤害后
    protected OnHurtAfter(msg:OnHurtBefore){}
    //该方法仅用于计算伤害.最终伤害由OnDamage结算.
    protected CalcDamageValue(msg:CalcDamageValueMessage){}
    //该方法仅用于计算伤害.最终伤害由OnDamage结算.
    protected CalcFinalDamageValue(msg:CalcFinalDamageValueMessage){}
    //造成伤害前
    protected OnDamageBefore(msg:OnDamageBefore){}
    //造成伤害后
    protected OnDamageAfter(msg:OnDamageAfter){}
    //回合结束时
    protected AtEndOfTurn(msg:AtEndOfTurnMessage){}
    //回合开始时
    protected AtStartOfTurn(msg:AtStartOfTurnMessage){}
    //死亡时
    protected OnDeath(msg:OnDeathMessage){}
    //回合结束和弃牌结束之间,怪物也会触发该函数
    AtEndOfTurnPreEndTurnCards(msg: AtEndOfTurnPreEndTurnCardsMessage) {}
    PreModifyBlock(msg:PreModifyBlockMessage) {}
    //
    PreUseCard(msg:OnPreUseCardMessage) {}

    protected OnModified(before:number,after:number){
        if(after == 0){
            this.AddToBot(new RemoveSpecificPowerAction(this.Owner,this.Owner,this));
        }
    }
}

export enum PowerType{
    Buff,
    Debuff,
    Special
}

File: Unity/Assets/Scripts/ModTheSpire/Power/AbstractPower.cs
using ModTheSpire.Dungeon;
using ModTheSpire.Effect.Other;
using ModTheSpire.Event;
using ModTheSpire.Event.Ids;
using ModTheSpire.Unit;

namespace ModTheSpire.Power
{
    public abstract class AbstractPower
    {
        public abstract string Id { get; }
        public abstract string Name { get; }

        /// 图标
        public abstract string Icon { get; }
        
        private int mAmount;

        /// 层数
        public int Amount
        {
            get => mAmount;
            set
            {
                var preAmount = mAmount;
                mAmount = value;
                OnModified(preAmount, value);
                DungeonManager.MessageManager.Send(new PowerModifiedAmountMessage(this, preAmount, value));
            }
        }

        //可否叠加
        public bool CanStack { get; set; }

        //可否驱散
        public bool CanRemove { get; set; }

        //结算优先级
        public int Priority { get; set; }

        //该Buff的拥有者
        public AbstractCreature Owner { get; set; }

        //该Buff的施法来源
        public AbstractCreature Source { get; set; }

        //能力类型(Buff,Debuff)
        public abstract PowerType Type { get; }
        
        ///获得描述
        public abstract string GetDescription(object[] args);

        //重载该方法需要调用super.OnInit();
        public virtual void OnInit()
        {
            DungeonManager.MessageManager.Add<OnHurtBeforeMessage>(OnHurtBefore);
            DungeonManager.MessageManager.Add<OnHurtMessage>(OnHurt);
            DungeonManager.MessageManager.Add<OnHurtAfterMessage>(OnHurtAfter);
            DungeonManager.MessageManager.Add<AtEndOfTurnMessage>(AtEndOfTurn);
            DungeonManager.MessageManager.Add<AtStartOfTurnMessage>(AtStartOfTurn);
            DungeonManager.MessageManager.Add<CalcDamageValueMessage>(CalcDamageValue);
            DungeonManager.MessageManager.Add<OnDamageBeforeMessage>(OnDamageBefore);
            DungeonManager.MessageManager.Add<OnDamageAfterMessage>(OnDamageAfter);
            DungeonManager.MessageManager.Add<CalcFinalDamageValueMessage>(CalcFinalDamageValue);
            DungeonManager.MessageManager.Add<CalcHurtValueMessage>(CalcHurtValue);
            DungeonManager.MessageManager.Add<CalcFinalHurtValueMessage>(CalcFinalHurtValue);
            DungeonManager.MessageManager.Add<OnDeathMessage>(OnDeath);
            DungeonManager.MessageManager.Add<AtEndOfTurnPreEndTurnCardsMessage>(AtEndOfTurnPreEndTurnCards);
            DungeonManager.MessageManager.Add<PreModifyBlockMessage>(PreModifyBlock);
        }

        //重载该方法需要调用super.OnRemove();
        public virtual void OnRemove()
        {
            DungeonManager.MessageManager.Remove<OnHurtBeforeMessage>(OnHurtBefore);
            DungeonManager.MessageManager.Remove<OnHurtMessage>(OnHurt);
            DungeonManager.MessageManager.Remove<OnHurtAfterMessage>(OnHurtAfter);
            DungeonManager.MessageManager.Remove<AtEndOfTurnMessage>(AtEndOfTurn);
            DungeonManager.MessageManager.Remove<AtStartOfTurnMessage>(AtStartOfTurn);
            DungeonManager.MessageManager.Remove<CalcDamageValueMessage>(CalcDamageValue);
            DungeonManager.MessageManager.Remove<OnDamageBeforeMessage>(OnDamageBefore);
            DungeonManager.MessageManager.Remove<OnDamageAfterMessage>(OnDamageAfter);
            DungeonManager.MessageManager.Remove<CalcFinalDamageValueMessage>(CalcFinalDamageValue);
            DungeonManager.MessageManager.Remove<CalcHurtValueMessage>(CalcHurtValue);
            DungeonManager.MessageManager.Remove<CalcFinalHurtValueMessage>(CalcFinalHurtValue);
            DungeonManager.MessageManager.Remove<OnDeathMessage>(OnDeath);
            DungeonManager.MessageManager.Remove<AtEndOfTurnPreEndTurnCardsMessage>(AtEndOfTurnPreEndTurnCards);
            DungeonManager.MessageManager.Remove<PreModifyBlockMessage>(PreModifyBlock);
        }

        public virtual void PlayApplyPowerSfx()
        {
        }

        protected void AddToBot(IGameAction action)
        {
            DungeonManager.ActionManager.AddToBottom(action);
        }

        protected void AddToTop(IGameAction action)
        {
            DungeonManager.ActionManager.AddToTop(action);
        }

        public void Flash(bool silent = true){
            //DungeonManager.EffectManager.Play(new GainPowerEffect(this, silent));
            //DungeonManager.EffectManager.Play(new FlashPowerEffect(this));
        }

        //该方法仅用于计算伤害.最终伤害由OnHurt结算.
        protected virtual void CalcHurtValue(CalcHurtValueMessage msg)
        {
        }

        //该方法仅用于计算伤害.最终伤害由OnHurt结算.
        protected virtual void CalcFinalHurtValue(CalcFinalHurtValueMessage msg)
        {
        }

        //当玩家受到伤害之前
        protected virtual void OnHurtBefore(OnHurtBeforeMessage msg)
        {
        }

        //当玩家受到伤害时
        protected virtual void OnHurt(OnHurtMessage msg)
        {
        }

        //当玩家受到伤害后
        protected virtual void OnHurtAfter(OnHurtAfterMessage msg)
        {
        }

        //该方法仅用于计算伤害.最终伤害由OnDamage结算.
        protected virtual void CalcDamageValue(CalcDamageValueMessage msg)
        {
        }

        //该方法仅用于计算伤害.最终伤害由OnDamage结算.
        protected virtual void CalcFinalDamageValue(CalcFinalDamageValueMessage msg)
        {
        }

        //造成伤害前
        protected virtual void OnDamageBefore(OnDamageBeforeMessage msg)
        {
        }

        //造成伤害后
        protected virtual void OnDamageAfter(OnDamageAfterMessage msg)
        {
        }

        //回合结束时
        protected virtual void AtEndOfTurn(AtEndOfTurnMessage msg)
        {
        }

        //回合开始时
        protected virtual void AtStartOfTurn(AtStartOfTurnMessage msg)
        {
        }

        //死亡时
        protected virtual void OnDeath(OnDeathMessage msg)
        {
        }

        //回合结束和弃牌结束之间,怪物也会触发该函数
        protected virtual void AtEndOfTurnPreEndTurnCards(AtEndOfTurnPreEndTurnCardsMessage msg)
        {
        }

        protected virtual void PreModifyBlock(PreModifyBlockMessage msg)
        {
        }

        //
        protected virtual void PreUseCard(OnUseCardAfterMessage msg)
        {
        }

        protected virtual void OnModified(int before,int after){
            if (after == 0)
            {
                this.AddToBot(new RemoveSpecificPowerAction(this.Owner, this.Owner, this));
            }
        }
    }
}